api/handler/multipart_upload.go

```diff
@@ -0,0 +1,536 @@
+		return
+	}
+
+	if err = api.EncodeToResponse(w, encodeListMultipartUploadsToResponse(list, p)); err != nil {
+		h.logAndSendError(w, "something went wrong", reqInfo, err)
+	}
+}
+
+func (h *handler) ListPartsHandler(w http.ResponseWriter, r *http.Request) {
+	reqInfo := api.GetReqInfo(r.Context())
+
+	bktInfo, err := h.obj.GetBucketInfo(r.Context(), reqInfo.BucketName)
+	if err != nil {
+		h.logAndSendError(w, "could not get bucket info", reqInfo, err)
+		return
+	}
+
+	if err := checkOwner(bktInfo, r.Header.Get(api.AmzExpectedBucketOwner)); err != nil {
+		h.logAndSendError(w, "expected owner doesn't match", reqInfo, err)
+		return
+	}
+
+	queryValues := reqInfo.URL.Query()
+	uploadID := queryValues.Get(uploadIdHeaderName)
+	additional := []zap.Field{zap.String("uploadID", uploadID), zap.String("Key", reqInfo.ObjectName)}
+
+	maxParts := maxSizePartsList
+	if queryValues.Get("max-parts") != "" {
+		val, err := strconv.Atoi(queryValues.Get("max-parts"))
+		if err != nil || val < 0 {
+			h.logAndSendError(w, "invalid MaxParts", reqInfo, errors.GetAPIError(errors.ErrInvalidMaxParts), additional...)
+			return
+		}
+		if val < maxSizePartsList {
+			maxParts = val
+		}
+	}
+
+	var partNumberMarker int
+	if queryValues.Get("part-number-marker") != "" {
+		if partNumberMarker, err = strconv.Atoi(queryValues.Get("part-number-marker")); err != nil || partNumberMarker <= 0 {
+			h.logAndSendError(w, "invalid PartNumberMarker", reqInfo, err, additional...)
+			return
+		}
+	}
+
+	p := &layer.ListPartsParams{
+		Info: &layer.UploadInfoParams{
+			UploadID: uploadID,
+			Bkt:      bktInfo,
+			Key:      reqInfo.ObjectName,
+		},
+		MaxParts:         maxParts,
+		PartNumberMarker: partNumberMarker,
+	}
+
+	list, err := h.obj.ListParts(r.Context(), p)
+	if err != nil {
+		h.logAndSendError(w, "could not list parts", reqInfo, err, additional...)
+		return
+	}
+
+	if err = api.EncodeToResponse(w, encodeListPartsToResponse(list, p)); err != nil {
+		h.logAndSendError(w, "something went wrong", reqInfo, err)
+	}
+}
+
+func (h *handler) AbortMultipartUploadHandler(w http.ResponseWriter, r *http.Request) {
+	reqInfo := api.GetReqInfo(r.Context())
+
+	bktInfo, err := h.obj.GetBucketInfo(r.Context(), reqInfo.BucketName)
+	if err != nil {
+		h.logAndSendError(w, "could not get bucket info", reqInfo, err)
+		return
+	}
+
+	if err := checkOwner(bktInfo, r.Header.Get(api.AmzExpectedBucketOwner)); err != nil {
+		h.logAndSendError(w, "expected owner doesn't match", reqInfo, err)
+		return
+	}
+
+	queryValues := reqInfo.URL.Query()
+	uploadID := queryValues.Get(uploadIdHeaderName)
+	additional := []zap.Field{zap.String("uploadID", uploadID), zap.String("Key", reqInfo.ObjectName)}
+
+	p := &layer.UploadInfoParams{
+		UploadID: uploadID,
+		Bkt:      bktInfo,
+		Key:      reqInfo.ObjectName,
+	}
+
+	info, err := h.obj.GetUploadInitInfo(r.Context(), p)
+	if err != nil {
+		h.logAndSendError(w, "could not get upload to abort", reqInfo, err, additional...)
+		return
+	}
```

Comment:
I think we should try abort multipart uploads even if the 0 part doesn't exist. It could be deleted by some reason while other parts still exist.